kubectl-volsync/cmd/migration_create.go

```diff
@@ -36,70 +91,319 @@ var migrationCreateCmd = &cobra.Command{
 	if err != nil {
 		return err
 	}
-	relation, err := createRelationship(configDir, rName, MigrationRelationship)
+	// Creates the Namespace if it doesn't exist
+	_, err = mc.ensureNamespace(ctx)
 	if err != nil {
 		return err
 	}
-
-	// Insert information into the relationship & save it
-	cap, err := cmd.Flags().GetString("capacity")
+	// Creates the PVC if it doesn't exist
+	_, err = mc.ensureDestPVC(ctx)
 	if err != nil {
 		return err
 	}
-	relation.Set("capacity", cap)
 
-	if err = relation.Save(); err != nil {
+	// Creates the RD if it doesn't exist
+	_, err = mc.ensureReplicationDestination(ctx)
+	if err != nil {
+		return err
+	}
+	// Save the destination details into relationship file
+	if err = mc.mr.Save(); err != nil {
 		return fmt.Errorf("unable to save relationship configuration: %w", err)
 	}
+	return nil
+}
+
+func (mc *migrationCreate) ensureNamespace(ctx context.Context) (*corev1.Namespace, error) {
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: mc.Namespace,
+		},
+	}
+	if err := mc.clientObject.Create(ctx, ns); err != nil {
+		if kerrs.IsAlreadyExists(err) {
+			klog.Infof("Namespace: \"%s\" is found, proceeding with the same",
+				mc.Namespace)
+			return ns, nil
+		}
+		return nil, err
+	}
+	klog.Infof("Created Destination Namespace: \"%s\" in Cluster: \"%s\"", ns.Name, ns.ClusterName)
 
-	// TODO: Set up objects on the cluster
-	_, _ = newClient("")
+	return ns, nil
+}
 
-	return nil
+func (mc *migrationCreate) ensureDestPVC(ctx context.Context) (*corev1.PersistentVolumeClaim, error) {
+	if mc.PVC == nil {
+		PVC, err := mc.createDestinationPVC(ctx)
+		if err != nil {
+			return nil, err
+		}
+		mc.PVC = PVC
+	} else {
+		klog.Infof("Destination PVC: \"%s\" is found in Namespace: \"%s\" and is used to create replication destination",
+			mc.PVC.Name, mc.PVC.Namespace)
+	}
+
+	return mc.PVC, nil
+}
+
+func (mc *migrationCreate) createDestinationPVC(ctx context.Context) (*corev1.PersistentVolumeClaim, error) {
+	destPVC := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      mc.PVCName,
+			Namespace: mc.Namespace,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes:      mc.AccessModes,
+			StorageClassName: mc.StorageClassName,
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: *mc.Capacity,
+				},
+			},
+		},
+	}
+
+	if err := mc.clientObject.Create(ctx, destPVC); err != nil {
+		return nil, err
+	}
+
+	klog.Infof("Created Destination PVC: \"%s\" in NameSpace: \"%s\" and Cluster: \"%s\" ",
+		destPVC.Name, destPVC.Namespace, destPVC.ClusterName)
+
+	return destPVC, nil
+}
+
+func (mc *migrationCreate) getDestinationPVC(ctx context.Context) *corev1.PersistentVolumeClaim {
+	destPVC := &corev1.PersistentVolumeClaim{}
+	pvcInfo := types.NamespacedName{
+		Namespace: mc.Namespace,
+		Name:      mc.PVCName,
+	}
+	err := mc.clientObject.Get(ctx, pvcInfo, destPVC)
+	if err != nil {
+		return nil
+	}
```

Comment:
This treats all errors the same. "not found" should be different from "access denied".
not found --> we need to create it
other error --> problem we need to report

But I'm not convinced you actually need this fn at all. You really just need an implementation of ensureDestPVC() that will Get() and if not found Create(). In both cases, returning (*PVC, error).

Reply:
The function `getDestinationPVC` is called at the start of `Run`. If PVC is not found and capacity is not provided, we ask the user to provide `capacity`. 
Then in the later part of `Run`, we do `ensureDestPVC()`. If we do `getDestinationPVC` as part of `ensureDestPVC` then we need to do some cleanup like delete ns etc as we come to `ensureDestPVC` after `ensureNamespace`.

Reply:
That's a good point. It is nice to not have to clean up the NS creation.